crates/net/eth-wire/src/types/broadcast.rs

```diff
@@ -529,103 +516,156 @@ impl HandleAnnouncement for NewPooledTransactionHashes66 {
-        for (i, hash) in self.0.iter().enumerate() {
-            if !f(hash) {
-                indices_to_remove.push(i);
-            }
-        }
+    fn dedup(self) -> PartiallyValidData<Self::Value> {
+        let Self(hashes) = self;
 
-        let mut removed_hashes = Vec::with_capacity(indices_to_remove.len());
+        let mut deduped_data = HashMap::with_capacity(hashes.len());
 
-        for index in indices_to_remove.into_iter().rev() {
-            let hash = self.0.remove(index);
-            removed_hashes.push(hash);
+        let noop_value: TxMetadata = None;
+
+        for hash in hashes.into_iter().rev() {
+            deduped_data.insert(hash, noop_value);
         }
 
-        Self(removed_hashes)
+        PartiallyValidData::from_raw_data_eth66(deduped_data)
     }
+}
 
-    fn msg_version(&self) -> EthVersion {
-        EthVersion::Eth66
-    }
+/// Interface for handling announcement data in filters in the transaction manager and transaction
+/// pool. Note: this trait may disappear when distinction between eth66 and eth68 hashes is more
+/// clearly defined, see <https://github.com/paradigmxyz/reth/issues/6148>.
+pub trait HandlePayload {
+    /// The announcement contains no entries.
+    fn is_empty(&self) -> bool;
+
+    /// Returns the number of entries.
+    fn len(&self) -> usize;
+
+    /// Retain only entries for which the hash in the entry satisfies a given predicate, return
+    /// the rest.
+    fn retain_by_hash(&mut self, f: impl FnMut(&TxHash) -> bool) -> Self;
 }
 
-/// Announcement data that has been validated according to the configured network. For an eth68
-/// announcement, values of the map are `Some((u8, usize))` - the tx metadata. For an eth66
-/// announcement, values of the map are `None`.
-#[derive(Debug, Deref, DerefMut, IntoIterator, Constructor)]
-pub struct ValidAnnouncementData {
+macro_rules! hash_map_handle_payload_impl {
+    ($data_ty:ty, $(<$generic:ident>)?) => {
+        impl$(<$generic>)? HandlePayload for $data_ty {
+            fn is_empty(&self) -> bool {
+                self.data.is_empty()
+            }
+
+            fn len(&self) -> usize {
+                self.data.len()
+            }
+
+            fn retain_by_hash(&mut self, mut f: impl FnMut(&TxHash) -> bool) -> Self {
+                let data = std::mem::take(&mut self.data);
+
+                let (keep, rest) = data.into_iter().partition(|(hash, _)| f(hash));
+
+                self.data = keep;
+
+                Self { data: rest, version: self.version }
+            }
+        }
+    };
+}
+
+/// Data that has passed an initial validation pass that is not specific to any mempool message
+/// type.
+#[derive(Debug, Deref, DerefMut, IntoIterator)]
+pub struct PartiallyValidData<V> {
     #[deref]
     #[deref_mut]
     #[into_iterator]
-    data: HashMap<TxHash, Option<(u8, usize)>>,
-    version: EthVersion,
+    data: HashMap<TxHash, V>,
+    version: Option<EthVersion>,
```

Comment:
why is this an option now?

Reply:
https://github.com/paradigmxyz/reth/pull/6439#discussion_r1493779489